pymmcore_plus/mda/_engine.py

```diff
@@ -0,0 +1,143 @@
+import time
+from abc import abstractmethod
+from typing import TYPE_CHECKING, Protocol, runtime_checkable
+
+from loguru import logger
+from useq import MDASequence
+
+from .events import PMDASignaler, _get_auto_MDA_callback_class
+
+if TYPE_CHECKING:
+    from ..core import CMMCorePlus
+
+
+@runtime_checkable
+class PMDAEngine(Protocol):
+    @property
+    @abstractmethod
+    def events(self) -> PMDASignaler:
+        """Return the MDA events object."""
+
+    @abstractmethod
+    def is_running(self) -> bool:
+        """Return whether currently running an Acquistion"""
+
+    @abstractmethod
+    def cancel(self):
+        """Cancel the MDA."""
+
+    @abstractmethod
+    def toggle_pause(self):
+        """Switch whether the MDA is paused."""
+
+    @abstractmethod
+    def is_paused(self) -> bool:
+        """Returns whether the acquistion is currently paused."""
+
+    @abstractmethod
+    def run(self, sequence: MDASequence):
+        """Start the acquisition loop blocking the current thread."""
+
+
+class MDAEngine(PMDAEngine):
+    def __init__(self, mmc: "CMMCorePlus" = None) -> None:
+        self._mmc = mmc
+        self._events = _get_auto_MDA_callback_class()()
+        self._canceled = False
+        self._paused = False
+        self._running = False
+
+    @property
+    def events(self) -> PMDASignaler:
+        return self._events
+
+    def is_running(self) -> bool:
+        return self._running
```

Comment:
only real heads up is that I added this so that we can make sure an MDA isn't running when the user registers a new engine. otherwise you can end up losing track of it and have no way to cancel it.